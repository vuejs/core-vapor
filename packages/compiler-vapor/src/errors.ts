import {
  CompilerError,
  SourceLocation,
  createCompilerError,
} from '@vue/compiler-dom'

export interface VaporCompilerError extends CompilerError {
  code: VaporErrorCodes
}

export function createVaporCompilerError(
  code: VaporErrorCodes,
  loc?: SourceLocation,
) {
  return createCompilerError(
    code,
    loc,
    __DEV__ || !__BROWSER__ ? VaporErrorMessages : undefined,
  ) as VaporCompilerError
}

export enum VaporErrorCodes {
  X_V_PLACEHOLDER = 100,
  __EXTEND_POINT__,
}

export const VaporErrorMessages: Record<VaporErrorCodes, string> = {
  [VaporErrorCodes.X_V_PLACEHOLDER]: `[placeholder]`,

  // just to fulfill types
  [VaporErrorCodes.__EXTEND_POINT__]: ``,
}
