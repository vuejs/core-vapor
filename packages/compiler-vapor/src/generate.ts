import {
  type CodegenOptions,
  type CodegenResult,
  type Position,
  type SourceLocation,
  NewlineType,
  advancePositionWithMutation,
  locStub,
  BindingTypes,
  createSimpleExpression,
  walkIdentifiers,
  advancePositionWithClone,
  isSimpleIdentifier,
} from '@vue/compiler-dom'
import {
  type IRDynamicChildren,
  type RootIRNode,
  type SetPropIRNode,
  type IRExpression,
  type OperationNode,
  type VaporHelper,
  type SetEventIRNode,
  type WithDirectiveIRNode,
  type SetTextIRNode,
  type SetHtmlIRNode,
  type CreateTextNodeIRNode,
  type InsertNodeIRNode,
  type PrependNodeIRNode,
  type AppendNodeIRNode,
  IRNodeTypes,
} from './ir'
import { SourceMapGenerator } from 'source-map-js'
import { camelize, isString, makeMap } from '@vue/shared'
import type { Identifier } from '@babel/types'

// remove when stable
// @ts-expect-error
function checkNever(x: never): never {}

export interface CodegenContext extends Required<CodegenOptions> {
  source: string
  code: string
  line: number
  column: number
  offset: number
  indentLevel: number
  map?: SourceMapGenerator

  push(
    code: string,
    newlineIndex?: number,
    loc?: SourceLocation,
    name?: string,
  ): void
  pushWithNewline(
    code: string,
    newlineIndex?: number,
    loc?: SourceLocation,
    name?: string,
  ): void
  indent(): void
  deindent(): void
  newline(): void

  helpers: Set<string>
  vaporHelpers: Set<string>
  helper(name: string): string
  vaporHelper(name: string): string
}

function createCodegenContext(
  ir: RootIRNode,
  {
    mode = 'function',
    prefixIdentifiers = mode === 'module',
    sourceMap = false,
    filename = `template.vue.html`,
    scopeId = null,
    optimizeImports = false,
    runtimeGlobalName = `Vue`,
    runtimeModuleName = `vue`,
    ssrRuntimeModuleName = 'vue/server-renderer',
    ssr = false,
    isTS = false,
    inSSR = false,
    inline = false,
    bindingMetadata = {},
  }: CodegenOptions,
) {
  const { helpers, vaporHelpers } = ir
  const context: CodegenContext = {
    mode,
    prefixIdentifiers,
    sourceMap,
    filename,
    scopeId,
    optimizeImports,
    runtimeGlobalName,
    runtimeModuleName,
    ssrRuntimeModuleName,
    ssr,
    isTS,
    inSSR,
    bindingMetadata,
    inline,

    source: ir.source,
    code: ``,
    column: 1,
    line: 1,
    offset: 0,
    indentLevel: 0,

    helpers,
    vaporHelpers,
    helper(name: string) {
      helpers.add(name)
      return `_${name}`
    },
    vaporHelper(name: VaporHelper) {
      vaporHelpers.add(name)
      return `_${name}`
    },
    push(code, newlineIndex = NewlineType.None, loc, name) {
      context.code += code
      if (!__BROWSER__ && context.map) {
        if (loc) addMapping(loc.start, name)

        if (newlineIndex === NewlineType.Unknown) {
          // multiple newlines, full iteration
          advancePositionWithMutation(context, code)
        } else {
          // fast paths
          context.offset += code.length
          if (newlineIndex === NewlineType.None) {
            // no newlines; fast path to avoid newline detection
            if (__TEST__ && code.includes('\n')) {
              throw new Error(
                `CodegenContext.push() called newlineIndex: none, but contains` +
                  `newlines: ${code.replace(/\n/g, '\\n')}`,
              )
            }
            context.column += code.length
          } else {
            // single newline at known index
            if (newlineIndex === NewlineType.End) {
              newlineIndex = code.length - 1
            }
            if (
              __TEST__ &&
              (code.charAt(newlineIndex) !== '\n' ||
                code.slice(0, newlineIndex).includes('\n') ||
                code.slice(newlineIndex + 1).includes('\n'))
            ) {
              throw new Error(
                `CodegenContext.push() called with newlineIndex: ${newlineIndex} ` +
                  `but does not conform: ${code.replace(/\n/g, '\\n')}`,
              )
            }
            context.line++
            context.column = code.length - newlineIndex
          }
        }
        if (loc && loc !== locStub) {
          addMapping(loc.end)
        }
      }
    },
    pushWithNewline(code, newlineIndex, node) {
      context.newline()
      context.push(code, newlineIndex, node)
    },
    indent() {
      ++context.indentLevel
    },
    deindent() {
      --context.indentLevel
    },
    newline() {
      newline(context.indentLevel)
    },
  }

  function newline(n: number) {
    context.push(`\n${`  `.repeat(n)}`, NewlineType.Start)
  }

  function addMapping(loc: Position, name: string | null = null) {
    // we use the private property to directly add the mapping
    // because the addMapping() implementation in source-map-js has a bunch of
    // unnecessary arg and validation checks that are pure overhead in our case.
    const { _names, _mappings } = context.map!
    if (name !== null && !_names.has(name)) _names.add(name)
    _mappings.add({
      originalLine: loc.line,
      originalColumn: loc.column - 1, // source-map column is 0 based
      generatedLine: context.line,
      generatedColumn: context.column - 1,
      source: filename,
      // @ts-ignore it is possible to be null
      name,
    })
  }

  if (!__BROWSER__ && sourceMap) {
    // lazy require source-map implementation, only in non-browser builds
    context.map = new SourceMapGenerator()
    context.map.setSourceContent(filename, context.source)
    context.map._sources.add(filename)
  }

  return context
}

// IR -> JS codegen
export function generate(
  ir: RootIRNode,
  options: CodegenOptions = {},
): CodegenResult {
  const ctx = createCodegenContext(ir, options)
  const { push, pushWithNewline, indent, deindent, newline } = ctx
  const { vaporHelper, helpers, vaporHelpers } = ctx

  const functionName = 'render'
  const isSetupInlined = !!options.inline
  if (isSetupInlined) {
    push(`(() => {`)
  } else {
    // placeholder for preamble
    newline()
    pushWithNewline(`export function ${functionName}(_ctx) {`)
  }
  indent()

  ir.template.forEach((template, i) => {
    if (template.type === IRNodeTypes.TEMPLATE_FACTORY) {
      // TODO source map?
      pushWithNewline(
        `const t${i} = ${vaporHelper('template')}(${JSON.stringify(
          template.template,
        )})`,
      )
    } else {
      // fragment
      pushWithNewline(
        `const t0 = ${vaporHelper('fragment')}()\n`,
        NewlineType.End,
      )
    }
  })

  {
    pushWithNewline(`const n${ir.dynamic.id} = t0()`)

    const children = genChildren(ir.dynamic.children)
    if (children) {
      pushWithNewline(
        `const ${children} = ${vaporHelper('children')}(n${ir.dynamic.id})`,
      )
    }

    for (const oper of ir.operation.filter(
      (oper): oper is WithDirectiveIRNode =>
        oper.type === IRNodeTypes.WITH_DIRECTIVE,
    )) {
      genWithDirective(oper, ctx)
    }

    for (const operation of ir.operation) {
      genOperation(operation, ctx)
    }

    for (const { operations } of ir.effect) {
      pushWithNewline(`${vaporHelper('effect')}(() => {`)
      indent()
      for (const operation of operations) {
        genOperation(operation, ctx)
      }
      deindent()
      pushWithNewline('})')
    }

    // TODO multiple-template
    // TODO return statement in IR
    pushWithNewline(`return n${ir.dynamic.id}`)
  }

  deindent()
  newline()
  if (isSetupInlined) {
    push('})()')
  } else {
    push('}')
  }

  let preamble = ''
  if (vaporHelpers.size)
    // TODO: extract import codegen
    preamble = `import { ${[...vaporHelpers]
      .map((h) => `${h} as _${h}`)
      .join(', ')} } from 'vue/vapor';`
  if (helpers.size)
    preamble = `import { ${[...helpers]
      .map((h) => `${h} as _${h}`)
      .join(', ')} } from 'vue';`

  if (!isSetupInlined) {
    ctx.code = preamble + ctx.code
  }

  return {
    code: ctx.code,
    ast: ir as any,
    preamble,
    map: ctx.map ? ctx.map.toJSON() : undefined,
  }
}

function genChildren(children: IRDynamicChildren) {
  let code = ''
  // TODO
  let offset = 0
  for (const [index, child] of Object.entries(children)) {
    const childrenLength = Object.keys(child.children).length
    if (child.ghost && child.placeholder === null && childrenLength === 0) {
      offset--
      continue
    }

    code += ` ${Number(index) + offset}: [`

    const id = child.ghost ? child.placeholder : child.id
    if (id !== null) code += `n${id}`

    const childrenString = childrenLength && genChildren(child.children)
    if (childrenString) code += `, ${childrenString}`

    code += '],'
  }

  if (!code) return ''
  return `{${code}}`
}

function genOperation(oper: OperationNode, context: CodegenContext) {
  // TODO: cache old value
  switch (oper.type) {
    case IRNodeTypes.SET_PROP:
      return genSetProp(oper, context)
    case IRNodeTypes.SET_TEXT:
      return genSetText(oper, context)
    case IRNodeTypes.SET_EVENT:
      return genSetEvent(oper, context)
    case IRNodeTypes.SET_HTML:
      return genSetHtml(oper, context)
    case IRNodeTypes.CREATE_TEXT_NODE:
      return genCreateTextNode(oper, context)
    case IRNodeTypes.INSERT_NODE:
      return genInsertNode(oper, context)
    case IRNodeTypes.PREPEND_NODE:
      return genPrependNode(oper, context)
    case IRNodeTypes.APPEND_NODE:
      return genAppendNode(oper, context)
    case IRNodeTypes.WITH_DIRECTIVE:
      // generated, skip
      return
    default:
      return checkNever(oper)
  }
}

function genSetProp(oper: SetPropIRNode, context: CodegenContext) {
  const { push, pushWithNewline, vaporHelper, helper } = context
  pushWithNewline(`${vaporHelper('setAttr')}(n${oper.element}, `)
  if (oper.runtimeCamelize) push(`${helper('camelize')}(`)
  genExpression(oper.key, context)
  if (oper.runtimeCamelize) push(`)`)
  push(`, undefined, `)
  genExpression(oper.value, context)
  push(')')
}

function genSetText(oper: SetTextIRNode, context: CodegenContext) {
  const { push, pushWithNewline, vaporHelper } = context
  pushWithNewline(`${vaporHelper('setText')}(n${oper.element}, undefined, `)
  genExpression(oper.value, context)
  push(')')
}

function genSetHtml(oper: SetHtmlIRNode, context: CodegenContext) {
  const { push, pushWithNewline, vaporHelper } = context
  pushWithNewline(`${vaporHelper('setHtml')}(n${oper.element}, undefined, `)
  genExpression(oper.value, context)
  push(')')
}

function genCreateTextNode(
  oper: CreateTextNodeIRNode,
  context: CodegenContext,
) {
  const { push, pushWithNewline, vaporHelper } = context
  pushWithNewline(`const n${oper.id} = ${vaporHelper('createTextNode')}(`)
  genExpression(oper.value, context)
  push(')')
}

function genInsertNode(oper: InsertNodeIRNode, context: CodegenContext) {
  const { pushWithNewline, vaporHelper } = context
  const elements = ([] as number[]).concat(oper.element)
  let element = elements.map((el) => `n${el}`).join(', ')
  if (elements.length > 1) element = `[${element}]`
  pushWithNewline(
    `${vaporHelper('insert')}(${element}, n${
      oper.parent
    }${`, n${oper.anchor}`})`,
  )
}

function genPrependNode(oper: PrependNodeIRNode, context: CodegenContext) {
  const { pushWithNewline, vaporHelper } = context
  pushWithNewline(
    `${vaporHelper('prepend')}(n${oper.parent}, ${oper.elements
      .map((el) => `n${el}`)
      .join(', ')})`,
  )
}

function genAppendNode(oper: AppendNodeIRNode, context: CodegenContext) {
  const { pushWithNewline, vaporHelper } = context
  pushWithNewline(
    `${vaporHelper('append')}(n${oper.parent}, ${oper.elements
      .map((el) => `n${el}`)
      .join(', ')})`,
  )
}

function genSetEvent(oper: SetEventIRNode, context: CodegenContext) {
  const { vaporHelper, push, pushWithNewline } = context

  pushWithNewline(`${vaporHelper('on')}(n${oper.element}, `)
  // second arg: event name
  genExpression(oper.key, context)
  push(', ')

  const { keys, nonKeys, options } = oper.modifiers
  if (keys.length) {
    push(`${vaporHelper('withKeys')}(`)
  }
  if (nonKeys.length) {
    push(`${vaporHelper('withModifiers')}(`)
  }

  // gen event handler
  push('(...args) => (')
  genExpression(oper.value, context)
  push(' && ')
  genExpression(oper.value, context)
  push('(...args))')

  if (nonKeys.length) {
    push(`, ${genArrayExpression(nonKeys)})`)
  }
  if (keys.length) {
    push(`, ${genArrayExpression(keys)})`)
  }
  if (options.length) {
    push(`, { ${options.map((v) => `${v}: true`).join(', ')} }`)
  }

  push(')')
}

function genWithDirective(oper: WithDirectiveIRNode, context: CodegenContext) {
  const { push, pushWithNewline, vaporHelper, bindingMetadata } = context
  const { dir } = oper

  // TODO merge directive for the same node
  pushWithNewline(`${vaporHelper('withDirectives')}(n${oper.element}, [[`)

  if (dir.name === 'show') {
    push(vaporHelper('vShow'))
  } else {
    const directiveReference = camelize(`v-${dir.name}`)
    // TODO resolve directive
    if (bindingMetadata[directiveReference]) {
      const directiveExpression = createSimpleExpression(directiveReference)
      directiveExpression.ast = null
      genExpression(directiveExpression, context)
    }
  }

  if (dir.exp) {
    push(', () => ')
    genExpression(dir.exp, context)
  } else if (dir.arg || dir.modifiers.length) {
    push(', void 0')
  }

  if (dir.arg) {
    push(', ')
    genExpression(dir.arg, context)
  } else if (dir.modifiers.length) {
    push(', void 0')
  }

  if (dir.modifiers.length) {
    push(', ')
    push('{ ')
    push(genDirectiveModifiers(dir.modifiers))
    push(' }')
  }
  push(']])')
  return
}

// TODO: other types (not only string)
function genArrayExpression(elements: string[]) {
  return `[${elements.map((it) => JSON.stringify(it)).join(', ')}]`
}

const isLiteralWhitelisted = /*#__PURE__*/ makeMap('true,false,null,this')

function genExpression(node: IRExpression, context: CodegenContext): void {
  const { push } = context
  if (isString(node)) return push(node)

  const { content: rawExpr, ast, isStatic, loc } = node
  if (__BROWSER__) {
    return push(rawExpr)
  }

  if (
    !context.prefixIdentifiers ||
    !node.content.trim() ||
    // there was a parsing error
    ast === false ||
    isLiteralWhitelisted(rawExpr)
  ) {
    return push(rawExpr, NewlineType.None, loc)
  }
  if (isStatic) {
    return push(JSON.stringify(rawExpr), NewlineType.None, loc)
  }

  if (ast === null) {
    // the expression is a simple identifier
    return genIdentifier(rawExpr, context, loc)
  }

  const ids: Identifier[] = []
  walkIdentifiers(
    ast!,
    (id) => {
      ids.push(id)
    },
    true,
  )
  if (ids.length) {
    ids.sort((a, b) => a.start! - b.start!)
    ids.forEach((id, i) => {
      // range is offset by -1 due to the wrapping parens when parsed
      const start = id.start! - 1
      const end = id.end! - 1
      const last = ids[i - 1]

      const leadingText = rawExpr.slice(last ? last.end! - 1 : 0, start)
      if (leadingText.length) push(leadingText, NewlineType.Unknown)

      const source = rawExpr.slice(start, end)
      genIdentifier(source, context, {
        start: advancePositionWithClone(node.loc.start, source, start),
        end: advancePositionWithClone(node.loc.start, source, end),
        source,
      })

      if (i === ids.length - 1 && end < rawExpr.length) {
        push(rawExpr.slice(end), NewlineType.Unknown)
      }
    })
  } else {
    push(rawExpr, NewlineType.Unknown)
  }
}

function genIdentifier(
  id: string,
  { inline, bindingMetadata, vaporHelper, push }: CodegenContext,
  loc?: SourceLocation,
): void {
  let name: string | undefined = id
  if (inline) {
    switch (bindingMetadata[id]) {
      case BindingTypes.SETUP_REF:
        name = id += '.value'
        break
      case BindingTypes.SETUP_MAYBE_REF:
        id = `${vaporHelper('unref')}(${id})`
        name = undefined
        break
    }
  } else {
    id = `_ctx.${id}`
  }
  push(id, NewlineType.None, loc, name)
}

function genDirectiveModifiers(modifiers: string[]) {
  return modifiers
    .map(
      (value) =>
        `${isSimpleIdentifier(value) ? value : JSON.stringify(value)}: true`,
    )
    .join(', ')
}
