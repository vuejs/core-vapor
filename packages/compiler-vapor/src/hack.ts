import type { Prettify } from '@vue/shared'
import type { DirectiveTransform, NodeTransform } from './transform'

type Overwrite<T, U> = Pick<T, Exclude<keyof T, keyof U>> &
  Pick<U, Extract<keyof U, keyof T>>

export type HackOptions<T> = Prettify<
  Overwrite<
    T,
    {
      nodeTransforms?: NodeTransform[]
      directiveTransforms?: Record<string, DirectiveTransform | undefined>
    }
  >
>
