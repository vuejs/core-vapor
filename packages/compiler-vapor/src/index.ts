export { parse } from '@vue/compiler-dom'
export { transform } from './transform'
export { generate } from './generate'
export { compile, type CompilerOptions } from './compile'
export * from './ir'
export * from './errors'
