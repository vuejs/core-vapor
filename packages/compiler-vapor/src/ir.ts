import type {
  CompoundExpressionNode,
  DirectiveNode,
  RootNode,
  SimpleExpressionNode,
  SourceLocation,
} from '@vue/compiler-dom'
import type { Prettify } from '@vue/shared'
import type { DirectiveTransform, NodeTransform } from './transform'

export enum IRNodeTypes {
  ROOT,
  TEMPLATE_FACTORY,
  FRAGMENT_FACTORY,

  SET_PROP,
  SET_TEXT,
  SET_EVENT,
  SET_HTML,

  INSERT_NODE,
  PREPEND_NODE,
  APPEND_NODE,
  CREATE_TEXT_NODE,

  WITH_DIRECTIVE,
}

export interface BaseIRNode {
  type: IRNodeTypes
  loc: SourceLocation
}

// TODO refactor
export type VaporHelper = keyof typeof import('../../runtime-vapor/src')

export interface RootIRNode extends BaseIRNode {
  type: IRNodeTypes.ROOT
  source: string
  node: RootNode
  template: Array<TemplateFactoryIRNode | FragmentFactoryIRNode>
  dynamic: IRDynamicInfo
  effect: IREffect[]
  operation: OperationNode[]
  helpers: Set<string>
  vaporHelpers: Set<VaporHelper>
}

export interface TemplateFactoryIRNode extends BaseIRNode {
  type: IRNodeTypes.TEMPLATE_FACTORY
  template: string
}

export interface FragmentFactoryIRNode extends BaseIRNode {
  type: IRNodeTypes.FRAGMENT_FACTORY
}

export interface SetPropIRNode extends BaseIRNode {
  type: IRNodeTypes.SET_PROP
  element: number
  key: IRExpression
  value: IRExpression
  runtimeCamelize: boolean
}

export interface SetTextIRNode extends BaseIRNode {
  type: IRNodeTypes.SET_TEXT
  element: number
  value: IRExpression
}

export interface SetEventIRNode extends BaseIRNode {
  type: IRNodeTypes.SET_EVENT
  element: number
  key: IRExpression
  value: IRExpression
  modifiers: {
    // modifiers for addEventListener() options, e.g. .passive & .capture
    options: string[]
    // modifiers that needs runtime guards, withKeys
    keys: string[]
    // modifiers that needs runtime guards, withModifiers
    nonKeys: string[]
  }
}

export interface SetHtmlIRNode extends BaseIRNode {
  type: IRNodeTypes.SET_HTML
  element: number
  value: IRExpression
}

export interface CreateTextNodeIRNode extends BaseIRNode {
  type: IRNodeTypes.CREATE_TEXT_NODE
  id: number
  value: IRExpression
}

export interface InsertNodeIRNode extends BaseIRNode {
  type: IRNodeTypes.INSERT_NODE
  element: number | number[]
  parent: number
  anchor: number
}

export interface PrependNodeIRNode extends BaseIRNode {
  type: IRNodeTypes.PREPEND_NODE
  elements: number[]
  parent: number
}

export interface AppendNodeIRNode extends BaseIRNode {
  type: IRNodeTypes.APPEND_NODE
  elements: number[]
  parent: number
}

export interface WithDirectiveIRNode extends BaseIRNode {
  type: IRNodeTypes.WITH_DIRECTIVE
  element: number
  dir: VaporDirectiveNode
}

export type IRNode =
  | OperationNode
  | RootIRNode
  | TemplateFactoryIRNode
  | FragmentFactoryIRNode
export type OperationNode =
  | SetPropIRNode
  | SetTextIRNode
  | SetEventIRNode
  | SetHtmlIRNode
  | CreateTextNodeIRNode
  | InsertNodeIRNode
  | PrependNodeIRNode
  | AppendNodeIRNode
  | WithDirectiveIRNode

export interface IRDynamicInfo {
  id: number | null
  referenced: boolean
  /** created by DOM API */
  ghost: boolean
  placeholder: number | null
  children: IRDynamicChildren
}
export type IRDynamicChildren = Record<number, IRDynamicInfo>

export type IRExpression = SimpleExpressionNode | string
export interface IREffect {
  // TODO multi-expression effect
  expressions: IRExpression[]
  operations: OperationNode[]
}

type Overwrite<T, U> = Pick<T, Exclude<keyof T, keyof U>> &
  Pick<U, Extract<keyof U, keyof T>>

export type HackOptions<T> = Prettify<
  Overwrite<
    T,
    {
      nodeTransforms?: NodeTransform[]
      directiveTransforms?: Record<string, DirectiveTransform | undefined>
    }
  >
>

export type VaporDirectiveNode = Overwrite<
  DirectiveNode,
  {
    exp: Exclude<DirectiveNode['exp'], CompoundExpressionNode>
    arg: Exclude<DirectiveNode['arg'], CompoundExpressionNode>
  }
>
