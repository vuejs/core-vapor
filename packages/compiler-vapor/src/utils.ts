import type { NumericLiteral, StringLiteral } from '@babel/types'
import { isGloballyAllowed } from '@vue/shared'
import {
  type AttributeNode,
  type ElementNode,
  NodeTypes,
  type SimpleExpressionNode,
  findDir as _findDir,
  findProp as _findProp,
  createSimpleExpression,
  isLiteralWhitelisted,
} from '@vue/compiler-dom'
import type { VaporDirectiveNode } from './ir'
import { EMPTY_EXPRESSION } from './transforms/utils'

export const findProp = _findProp as (
  node: ElementNode,
  name: string,
  dynamicOnly?: boolean,
  allowEmpty?: boolean,
) => AttributeNode | VaporDirectiveNode | undefined

/** find directive */
export const findDir = _findDir as (
  node: ElementNode,
  name: string | RegExp,
  allowEmpty?: boolean,
) => VaporDirectiveNode | undefined

export function propToExpression(prop: AttributeNode | VaporDirectiveNode) {
  return prop.type === NodeTypes.ATTRIBUTE
    ? prop.value
      ? createSimpleExpression(prop.value.content, true, prop.value.loc)
      : EMPTY_EXPRESSION
    : prop.exp
}

export function isConstantExpression(exp: SimpleExpressionNode) {
  return (
    isLiteralWhitelisted(exp.content) ||
    isGloballyAllowed(exp.content) ||
    getLiteralExpressionValue(exp) !== null
  )
}

export function resolveExpression(exp: SimpleExpressionNode) {
  if (!exp.isStatic) {
    const value = getLiteralExpressionValue(exp)
    if (value !== null) {
      return createSimpleExpression('' + value, true, exp.loc)
    }
  }
  return exp
}

export function getLiteralExpressionValue(
  exp: SimpleExpressionNode,
): number | string | null {
  if (
    !__BROWSER__ &&
    exp.ast &&
    ['StringLiteral', 'NumericLiteral'].includes(exp.ast.type)
  ) {
    return (exp.ast as StringLiteral | NumericLiteral).value
  }
  return exp.isStatic ? exp.content : null
}
