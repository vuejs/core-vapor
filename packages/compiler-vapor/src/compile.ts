import {
  type CompilerOptions as BaseCompilerOptions,
  ErrorCodes,
  type RootNode,
  createCompilerError,
  defaultOnError,
  parse,
} from '@vue/compiler-dom'
import { extend, isString } from '@vue/shared'
import {
  type DirectiveTransform,
  type NodeTransform,
  transform,
} from './transform'
import { type VaporCodegenResult, generate } from './generate'
import { transformOnce } from './transforms/vOnce'
import { transformElement } from './transforms/transformElement'
import { transformVHtml } from './transforms/vHtml'
import { transformVText } from './transforms/vText'
import { transformVBind } from './transforms/vBind'
import { transformVOn } from './transforms/vOn'
import { transformVShow } from './transforms/vShow'
import { transformRef } from './transforms/transformRef'
import { transformInterpolation } from './transforms/transformInterpolation'
import type { HackOptions } from './ir'
import { transformVModel } from './transforms/vModel'
import { transformVIf } from './transforms/vIf'
import { transformVFor } from './transforms/vFor'

export type CompilerOptions = HackOptions<BaseCompilerOptions>

// TODO: copied from @vue/compiler-core
// code/AST -> IR -> JS codegen
export function compile(
  source: string | RootNode,
  options: CompilerOptions = {},
): VaporCodegenResult {
  const onError = options.onError || defaultOnError
  const isModuleMode = options.mode === 'module'
  /* istanbul ignore if */
  if (__BROWSER__) {
    if (options.prefixIdentifiers === true) {
      onError(createCompilerError(ErrorCodes.X_PREFIX_ID_NOT_SUPPORTED))
    } else if (isModuleMode) {
      onError(createCompilerError(ErrorCodes.X_MODULE_MODE_NOT_SUPPORTED))
    }
  }

  const prefixIdentifiers =
    !__BROWSER__ && (options.prefixIdentifiers === true || isModuleMode)

  // TODO scope id
  // if (options.scopeId && !isModuleMode) {
  //   onError(createCompilerError(ErrorCodes.X_SCOPE_ID_NOT_SUPPORTED))
  // }

  const resolvedOptions = extend({}, options, {
    prefixIdentifiers,
  })
  const ast = isString(source) ? parse(source, resolvedOptions) : source
  const [nodeTransforms, directiveTransforms] =
    getBaseTransformPreset(prefixIdentifiers)

  if (!__BROWSER__ && options.isTS) {
    const { expressionPlugins } = options
    if (!expressionPlugins || !expressionPlugins.includes('typescript')) {
      resolvedOptions.expressionPlugins = [
        ...(expressionPlugins || []),
        'typescript',
      ]
    }
  }

  const ir = transform(
    ast,
    extend({}, resolvedOptions, {
      nodeTransforms: [
        ...nodeTransforms,
        ...(options.nodeTransforms || []), // user transforms
      ],
      directiveTransforms: extend(
        {},
        directiveTransforms,
        options.directiveTransforms || {}, // user transforms
      ),
    }),
  )

  return generate(ir, resolvedOptions)
}

export type TransformPreset = [
  NodeTransform[],
  Record<string, DirectiveTransform>,
]

export function getBaseTransformPreset(
  prefixIdentifiers?: boolean,
): TransformPreset {
  return [
    [
      transformOnce,
      transformRef,
      transformInterpolation,
      transformVIf,
      transformVFor,
      transformElement,
    ],
    {
      bind: transformVBind,
      on: transformVOn,
      html: transformVHtml,
      text: transformVText,
      show: transformVShow,
      model: transformVModel,
    },
  ]
}
