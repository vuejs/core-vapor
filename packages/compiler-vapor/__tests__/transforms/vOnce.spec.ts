import { BindingTypes, NodeTypes } from '@vue/compiler-dom'
import { IRNodeTypes } from '../../src'
import { getBaseTransformPreset } from '../../src/compile'
import { makeCompile } from './_utils'

const [nodeTransforms, directiveTransforms] = getBaseTransformPreset(true)
const compileWithOnce = makeCompile({
  nodeTransforms,
  directiveTransforms,
})

describe('compiler: v-once', () => {
  test('basic', () => {
    const { ir, code, helpers } = compileWithOnce(
      `<div v-once>
        {{ msg }}
        <span :class="clz" />
      </div>`,
      {
        bindingMetadata: {
          msg: BindingTypes.SETUP_REF,
          clz: BindingTypes.SETUP_REF,
        },
      },
    )

    expect(code).toMatchSnapshot()
    expect(helpers).lengthOf(0)
    expect(ir.effect).lengthOf(0)
    expect(ir.operation).toMatchObject([
      {
        id: 1,
        type: IRNodeTypes.CREATE_TEXT_NODE,
      },
      {
        element: 1,
        type: IRNodeTypes.SET_TEXT,
        value: {
          type: NodeTypes.SIMPLE_EXPRESSION,
          content: 'msg',
          isStatic: false,
        },
      },
      {
        element: 2,
        key: {
          type: NodeTypes.SIMPLE_EXPRESSION,
          content: 'class',
          isStatic: true,
        },
        type: IRNodeTypes.SET_PROP,
        value: {
          type: NodeTypes.SIMPLE_EXPRESSION,
          content: 'clz',
          isStatic: false,
        },
      },
      {
        type: IRNodeTypes.PREPEND_NODE,
        elements: [1],
        parent: 3,
      },
    ])
  })

  test('as root node', () => {
    const { ir, code, helpers } = compileWithOnce(`<div :id="foo" v-once />`)

    expect(code).toMatchSnapshot()
    expect(helpers).lengthOf(0)
    expect(ir.effect).lengthOf(0)
    expect(ir.operation).toMatchObject([
      {
        type: IRNodeTypes.SET_PROP,
        element: 1,
        key: {
          type: NodeTypes.SIMPLE_EXPRESSION,
          content: 'id',
          isStatic: true,
        },
        value: {
          type: NodeTypes.SIMPLE_EXPRESSION,
          content: 'foo',
          isStatic: false,
        },
      },
    ])
    expect(code).not.contains('effect')
  })

  test('on nested plain element', () => {
    const { ir, code, helpers } = compileWithOnce(
      `<div><div :id="foo" v-once /></div>`,
    )

    expect(code).toMatchSnapshot()
    expect(helpers).lengthOf(0)
    expect(ir.effect).lengthOf(0)
    expect(ir.operation).toMatchObject([
      {
        type: IRNodeTypes.SET_PROP,
        element: 1,
        runtimeCamelize: false,
        key: {
          type: NodeTypes.SIMPLE_EXPRESSION,
          content: 'id',
          isStatic: true,
        },
        value: {
          type: NodeTypes.SIMPLE_EXPRESSION,
          content: 'foo',
          isStatic: false,
        },
      },
    ])
  })

  test.todo('on component')
  test.todo('on slot outlet')

  test('inside v-once', () => {
    const { ir, code, helpers } = compileWithOnce(
      `<div v-once><div v-once/></div>`,
    )

    expect(code).toMatchSnapshot()
    expect(helpers).lengthOf(0)
    expect(ir.effect).lengthOf(0)
    expect(ir.operation).lengthOf(0)
  })

  test.todo('with hoistStatic: true')
  test.todo('with v-if/else')
  test.todo('with v-for')
})
