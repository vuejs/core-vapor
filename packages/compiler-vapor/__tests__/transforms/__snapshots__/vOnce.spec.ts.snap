// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`compiler: v-once > as root node 1`] = `
"import { template as _template, children as _children, setDynamicProp as _setDynamicProp } from 'vue/vapor';

export function render(_ctx) {
  const t0 = _template("<div></div>")
  const n0 = t0()
  const { 0: [n1],} = _children(n0)
  _setDynamicProp(n1, "id", undefined, _ctx.foo)
  return n0
}"
`;

exports[`compiler: v-once > basic 1`] = `
"import { template as _template, children as _children, createTextNode as _createTextNode, setText as _setText, setClass as _setClass, prepend as _prepend } from 'vue/vapor';

export function render(_ctx) {
  const t0 = _template("<div> <span></span></div>")
  const n0 = t0()
  const { 0: [n3, { 1: [n2],}],} = _children(n0)
  const n1 = _createTextNode(_ctx.msg)
  _setText(n1, undefined, _ctx.msg)
  _setClass(n2, "class", undefined, _ctx.clz)
  _prepend(n3, n1)
  return n0
}"
`;

exports[`compiler: v-once > inside v-once 1`] = `
"import { template as _template } from 'vue/vapor';

export function render(_ctx) {
  const t0 = _template("<div><div></div></div>")
  const n0 = t0()
  return n0
}"
`;

exports[`compiler: v-once > on nested plain element 1`] = `
"import { template as _template, children as _children, setDynamicProp as _setDynamicProp } from 'vue/vapor';

export function render(_ctx) {
  const t0 = _template("<div><div></div></div>")
  const n0 = t0()
  const { 0: [, { 0: [n1],}],} = _children(n0)
  _setDynamicProp(n1, "id", undefined, _ctx.foo)
  return n0
}"
`;
